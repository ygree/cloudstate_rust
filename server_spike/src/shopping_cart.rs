
use ::command_macro_derive::CommandDecoder;
use protocols::example::shoppingcart::{AddLineItem, RemoveLineItem, GetShoppingCart, persistence};
use protocols::example::shoppingcart::persistence::{Cart, ItemAdded, ItemRemoved};
use super::{EventSourcedEntity, CommandDecoder};
use ::prost::Message;
use bytes::Bytes;
use crate::HandleCommandContext;

#[derive(Default)]
pub struct ShoppingCartEntity(Cart);
//TODO use more convenient type for internal state, e.g. HashMap

//TODO consider using Attribute-like macros, e.g. #[commands(AddLineItem, RemoveLineItem, GetShoppingCart)]
// Combine command into one type.
#[derive(CommandDecoder)]
pub enum ShoppingCartCommand {
    AddLineItem(AddLineItem),
    RemoveLineItem(RemoveLineItem),
    GetShoppingCart(GetShoppingCart),
}

pub enum ShoppingCartEvent {
    ItemAdded(ItemAdded),
    ItemRemoved(ItemRemoved),
}

// impl ShoppingCartEvent {
//     fn from_message(msg: impl ::prost::Message) -> Option<ShoppingCartEvent> {
//         match msg {
//             cmd @ ItemAdded {..} => Some(cmd),
//             cmd @ ItemRemoved {..} => Some(cmd),
//             _ => None,
//         }
//     }
// }

impl EventSourcedEntity for ShoppingCartEntity {

    type Snapshot = Cart;
    type Command = ShoppingCartCommand;
    type Event = ShoppingCartEvent;

    fn snapshot_loaded(&mut self, snapshot: Self::Snapshot) {
        self.0 = snapshot;
        println!("Snapshot Loaded: {:?}", self.0);
    }

    fn handle_command(&self, command: Self::Command, context: &mut impl HandleCommandContext<Event=Self::Event>) {
        match command {
            ShoppingCartCommand::AddLineItem(add_line_item) => {
                println!("Handle command: {:?}", add_line_item);
                context.emit_event(
                    //TODO looks like too much boilerplate
                    ShoppingCartEvent::ItemAdded(
                        ItemAdded {
                            item: Some(
                                persistence::LineItem {
                                    product_id: add_line_item.product_id,
                                    name: add_line_item.name,
                                    quantity: add_line_item.quantity,
                                }
                            )
                        }
                    )
                );
            },
            ShoppingCartCommand::RemoveLineItem(remove_line_item) => {
                println!("Handle command: {:?}", remove_line_item);
            },
            ShoppingCartCommand::GetShoppingCart(get_shopping_cart) => {
                println!("Handle command: {:?}", get_shopping_cart);
            },
        }
    }

    fn handle_event(&mut self, event: Self::Event) {
        match event {
            ShoppingCartEvent::ItemAdded(item_added) => {
                println!("Handle event: {:?}", item_added);
                if let Some(item) = item_added.item {
                    self.0.items.push(item);
                }
            },
            ShoppingCartEvent::ItemRemoved(item_removed) => {
                println!("Handle event: {:?}", item_removed);
                //TODO remove item
            },
        }
    }

}

