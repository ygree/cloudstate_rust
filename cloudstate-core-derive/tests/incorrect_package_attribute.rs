use cloudstate_core_derive::AnyMessage;

mod shopping_cart;
use shopping_cart::*;

#[package("com.example.shoppingcart")]
#[derive(AnyMessage)]
pub enum ShoppingCartCommand {
    AddLine(AddLineItem),
    RemoveLine(RemoveLineItem),
    GetCart(GetShoppingCart),
}

fn main() {}