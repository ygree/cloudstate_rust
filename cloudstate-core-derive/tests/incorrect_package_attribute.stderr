warning: associated type `handleStream` should have an upper camel case name
   --> $DIR/cloudstate.eventsourced.rs:181:14
    |
181 |         type handleStream: Stream<Item = Result<super::EventSourcedStreamOut, tonic::Status>>
    |              ^^^^^^^^^^^^ help: convert the identifier to upper camel case: `HandleStream`
    |
    = note: `#[warn(non_camel_case_types)]` on by default

error: expected `=`
 --> $DIR/incorrect_package_attribute.rs:6:1
  |
6 | #[package("com.example.shoppingcart")]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
